### ISSUE TITLE
Parse error on HP ProDesk 600 G3 DM

### ISSUE THREAD

**Issue by simulacra75** [SCConfigMgr/ConfigMgr#120]
Error message is;

`Compare-BIOSVersion : Exception calling "Parse" with "1" argument(s): "Input string was not in a correct format."`

If i query the SMBIOSBIOSVERSION attribute it returns,  **P22 Ver. 02.22**

**Comment by LionZum** [comment]
```
				"*ver*" {
                    $CurrentBIOSVersion = [System.Version]::Parse(((($CurrentBIOSProperties.SMBIOSBIOSVersion).Trim($CurrentBIOSProperties.SMBIOSBIOSVersion.Split(".")[0])).TrimStart(".")).Trim())
				}
```

P21 Ver. 02.22 is getting trimmed to ". 02." right after the first trim

As a work around for now, I have changed
`$CurrentBIOSVersion = [System.Version]::Parse(((($CurrentBIOSProperties.SMBIOSBIOSVersion).Trim($CurrentBIOSProperties.SMBIOSBIOSVersion.Split(".")[0])).TrimStart(".")).Trim())`

to 

`$CurrentBIOSVersion = [System.Version]::Parse((((($CurrentBIOSProperties.SMBIOSBIOSVersion).Trim($CurrentBIOSProperties.SMBIOSBIOSVersion.Split(".")[0])).TrimStart(".")).Trim())+($CurrentBIOSProperties.SMBIOSBIOSVersion.Split(".")[2]))`

**Comment by NickolajA** [comment]
You're right, it does. But your code won't work for the latest BIOS versioning standard. Here's an example:

`$CurrentBIOSVersion = "P21 Ver. 02.22"
$CurrentBIOSVersion = "Q71 Ver. 01.06.03"

# Detect new versus old BIOS formats
switch -wildcard ($($CurrentBIOSVersion)) {
    "*ver*" {
        $CurrentBIOSVersion1 = [System.Version]::Parse(((($CurrentBIOSVersion).Trim($CurrentBIOSVersion.Split(".")[0])).TrimStart(".")).Trim())
        $CurrentBIOSVersion2 = [System.Version]::Parse((((($CurrentBIOSVersion).Trim($CurrentBIOSVersion.Split(".")[0])).TrimStart(".")).Trim())+($CurrentBIOSVersion.Split(".")[2]))
    }
    default {
        $CurrentBIOSVersion1 = "$($CurrentBIOSProperties.SystemBiosMajorVersion).$($CurrentBIOSProperties.SystemBiosMinorVersion)"
    }
}
$CurrentBIOSVersion1
$CurrentBIOSVersion2`

We'll look into this and see if we can come up with a better solution that works in both situations.

**Comment by NickolajA** [comment]
I believe this code should work for both scenarios:

`
($CurrentBIOSVersion).TrimStart($CurrentBIOSVersion.Split(".")[0]).TrimStart(".").Trim()
`

**Comment by NickolajA** [comment]
It's been changed in 2.0.8 of the script, could you please verify if it works? Thanks.

**Comment by LionZum** [comment]
I’ll give it a whirl tomorrow at work and let you know.

**Comment by fl4tout** [comment]
Still getting this in 2.0.9.  

![image](https://user-images.githubusercontent.com/9938212/57150069-dc49fd00-6d92-11e9-9b0a-9cfc3d510d61.png)

**Comment by NickolajA** [comment]
Ok that's can easily be explained since your model contains chars after the version number and not only before. What model is that from?

**Comment by fl4tout** [comment]
This particular one is a Hewlett-Packard ProDesk 600 G4 Small Form Factor PC (Models included:83ee).

**Comment by NickolajA** [comment]
This should now be fixed in 2.1.0, could you please give it a try?

**Comment by fl4tout** [comment]
Tried with 2.1.0 on a device a few minutes ago.  SMSTS.log indicates its error is the same but on line 549 now.

**Comment by NickolajA** [comment]
What if you run the following in PowerShell on that model:

```
$CurrentBIOSVersion = Get-WmiObject -Class Win32_BIOS | Select-Object -ExpandProperty SMBIOSBIOSVersion
[System.Version]::Parse(($CurrentBIOSVersion).TrimStart($CurrentBIOSVersion.Split(".")[0]).TrimStart(".").Trim().Split(" ")[0])
```

**Comment by fl4tout** [comment]
That returns the following: 
![mbm](https://user-images.githubusercontent.com/9938212/57304747-84b6d480-70a5-11e9-9ff0-996f5eff5414.png)

**Comment by NickolajA** [comment]
Is this on another machine? You had 02.06.03 previously in your screenshot. Could you extract the SMBIOSBIOSVersion property value please. Also, what if you run main part of the trimming itself, like the following:

```
$CurrentBIOSProperties = (Get-WmiObject -Class Win32_BIOS | Select-Object -Property *)

# Detect new versus old BIOS formats
switch -wildcard ($($CurrentBIOSProperties.SMBIOSBIOSVersion)) {
    "*ver*" {
        if ($CurrentBIOSProperties.SMBIOSBIOSVersion -match '.F.\d+$') {
            $CurrentBIOSVersion = ($CurrentBIOSProperties.SMBIOSBIOSVersion -split "Ver.")[1].Trim()
            $BIOSVersionParseable = $false
        }
        else {
            $CurrentBIOSVersion = [System.Version]::Parse(($CurrentBIOSProperties.SMBIOSBIOSVersion).TrimStart($CurrentBIOSProperties.SMBIOSBIOSVersion.Split(".")[0]).TrimStart(".").Trim().Split(" ")[0])
            $BIOSVersionParseable = $true
        }
    }
    default {
        $CurrentBIOSVersion = "$($CurrentBIOSProperties.SystemBiosMajorVersion).$($CurrentBIOSProperties.SystemBiosMinorVersion)"
        $BIOSVersionParseable = $true
    }
}
$CurrentBIOSVersion
```

What do you get?

**Comment by fl4tout** [comment]
This is the same machine.  You bring up an interesting point, though.  The error in smsts.log references "02.06.03 A 1" I believe that is the available BIOS version.  The ApplyBIOSPackage.log lists the following for version matching:
"Current BIOS version determined as: Q07 Ver. 02.02.04" (At the top of the log)
"Current BIOS release detected as 2.2.4" (Bottom of the log)
"Available BIOS release detected as 02.06.03 A 1." (Bottom of the log)

I ran the code snippet as well as got the SMBIOSBIOSVersion for you here:
![image](https://user-images.githubusercontent.com/9938212/57306517-951c7e80-70a8-11e9-9508-3b062614679d.png)

**Comment by NickolajA** [comment]
Ah it's the available version from the BIOS package properties, that makes sense it's just another part of the code that needs to be trimmed then.

**Comment by NickolajA** [comment]
Updated 2.1.0 with a small fix, please try again.

**Comment by NickolajA** [comment]
That's a good catch. It has been addressed in 1.0.5. Thanks
